### PR TITLE
CORE: Do a bit of cleanup of core fetching

### PR DESCRIPTION
**Pull request by levitte**
Some data, like the library context, were passed both through higher
level callback structures and through arguments to those same higher
level callbacks.  This is a bit unnecessary, so we rearrange the
callback arguments to simply pass that callback structure and rely on
the higher level fetching functionality to pick out what data they
need from that structure.

-----

Only the last commit is part of this PR.  The rest comes from #15604 and #15737

### CODE REVIEW

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
LGTM, I'll recheck after the #15604 is merged and this is rebased.

**Comment by levitte** [comment]
Now that #15604 is in, this is in truth only one commit

**Comment by levitte** [comment]
@t8m, if you want to do your recheck, now would be a good time :wink:

**Review by t8m — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Pull request closed without merging**
